### PR TITLE
Added reusable admin view

### LINKED ISSUE #17: Admin reusables view
As an admin I'd like to edit reusable materials dropdown on admin view

### PR DESCRIPTION
**Pull request by EVP1989**
Resolves #17

### CODE REVIEW

File: src/components/admin/reusables.tsx [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+    try {
+      const fetchedMaterials = await Api.getReusableMaterialApi(keycloak.token).listReusableMaterials();
+      setMaterials(fetchedMaterials);
+      setLoading(false);
+    } catch (error) {
+      errorContext.setError(strings.errorHandling.materials.list, error);
+      setLoading(false);
+    }
+  };
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:45]:
```suggestion
  /**
   * Fetches list of reusable materials and building parts
   */
  const fetchReusableMaterials = async () => {
    if (!keycloak?.token) {
      return;
    }

    setLoading(true);

    try {
      const fetchedMaterials = await Api.getReusableMaterialApi(keycloak.token).listReusableMaterials();
      setMaterials(fetchedMaterials);
    } catch (error) {
      errorContext.setError(strings.errorHandling.materials.list, error);
    }

    setLoading(false);
  };
```

File: src/components/admin/reusables.tsx [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+import { selectKeycloak } from "features/auth-slice";
+import { ErrorContext } from "components/error-handler/error-handler";
+import { ReusableMaterial } from "generated/client";
+
+/**
+ * Component for reusable materials dropdown menu editor
+ */
+const Reusables: React.FC = () => {
+  const errorContext = React.useContext(ErrorContext);
+  const keycloak = useAppSelector(selectKeycloak);
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:18]:
```suggestion
  const errorContext = React.useContext(ErrorContext);
  const keycloak = useAppSelector(selectKeycloak);

```

File: src/components/admin/reusables.tsx [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+          name: newMaterialName,
+          metadata: {}
+        }
+      });
+      setMaterials([ ...materials, createdReusableMaterial ]);
+      setNewMaterialName(undefined);
+    } catch (error) {
+      errorContext.setError(strings.errorHandling.materials.create, error);
+    }
+    setAddingMaterial(false);
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:74]:
```suggestion
    }

    setAddingMaterial(false);
```

File: src/components/admin/reusables.tsx [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+      const deletableIndex = materials.findIndex(material => material.id === deletableMaterial?.id);
+      const newReusableMaterials = [ ...materials ];
+      newReusableMaterials.splice(deletableIndex, 1);
+      setMaterials(newReusableMaterials);
+      setDeletableMaterial(undefined);
+    } catch (error) {
+      errorContext.setError(strings.errorHandling.materials.delete, error);
+    }
+    setDeletingMaterial(false);
+  };
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:96]:
```suggestion
  /**
   * Event handler for deleting reusable material confirm
   */
  const onDeleteReusableMaterialConfirm = async () => {
    if (!keycloak?.token || !deletableMaterial?.id) {
      return;
    }

    try {
      await Api.getReusableMaterialApi(keycloak.token).deleteReusableMaterial({ reusableMaterialId: deletableMaterial.id });
      setMaterials(materials.filter(material => material.id !== deletableMaterial.id));
      setDeletableMaterial(undefined);
    } catch (error) {
      errorContext.setError(strings.errorHandling.materials.delete, error);
    }

    setDeletingMaterial(false);
  };
```

File: src/components/admin/reusables.tsx [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+      const editableIndex = materials.findIndex(material => material.id === editableMaterial?.id);
+      const newReusableMaterials = [ ...materials ];
+      newReusableMaterials[editableIndex] = editableMaterial;
+      setMaterials(newReusableMaterials);
+      setEditableMaterial(undefined);
+    } catch (error) {
+      errorContext.setError(strings.errorHandling.materials.update, error);
+    }
+    setEditingMaterial(false);
+  };
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:119]:
```suggestion
  /**
   * Event handler for editing reusable material confirm
   */
  const onEditReusableMaterialConfirm = async () => {
    if (!keycloak?.token || !editableMaterial?.id) {
      return;
    }

    try {
      const updatedMaterial = await Api.getReusableMaterialApi(keycloak.token).updateReusableMaterial({
        reusableMaterialId: editableMaterial.id, reusableMaterial: editableMaterial
      });

      setMaterials(materials.map(material => (material.id === updatedMaterial.id ? updatedMaterial : material)));
      setEditableMaterial(undefined);
    } catch (error) {
      errorContext.setError(strings.errorHandling.materials.update, error);
    }

    setEditingMaterial(false);
  };
```

File: src/components/admin/reusables.tsx [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+  /**
+   * Even handler for editable material name change
+   */
+  const handleEditableNameChange: React.ChangeEventHandler<HTMLTextAreaElement | HTMLInputElement> = ({ target }) => {
+    const { value } = target;
+    if (editableMaterial) {
+      setEditableMaterial({ ...editableMaterial, name: value });
+    }
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:148]:
```suggestion
    if (editableMaterial) {
      setEditableMaterial({ ...editableMaterial, name: target.value });
    }
```

File: src/components/admin/reusables.tsx [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    setEditingMaterial(false);
+  };
+
+  /**
+   * Event handler delete icon click
+   *
+   * @param material material
+   */
+  const deleteIconClick = (material : ReusableMaterial) => {
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:126]:
```suggestion
  const deleteIconClick = (material : ReusableMaterial) => () => {
```

File: src/components/admin/reusables.tsx [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+    setDeletableMaterial(material);
+    setDeletingMaterial(true);
+  };
+
+  /**
+   * Event handler edit icon click
+   *
+   * @param material material
+   */
+  const editIconClick = (material : ReusableMaterial) => {
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:136]:
```suggestion
  const editIconClick = (material : ReusableMaterial) => () => {
```

File: src/components/admin/reusables.tsx [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+      <MaterialItem key={ material.id }>
+        <MaterialText primary={ material.name }/>
+        <ListItemSecondaryAction>
+          <IconButton onClick={ () => deleteIconClick(material) }>
+            <Delete/>
+          </IconButton>
+          <IconButton onClick={ () => editIconClick(material) }>
+            <Edit/>
+          </IconButton>
+        </ListItemSecondaryAction>
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:167]:
```suggestion
        <ListItemSecondaryAction>
          <IconButton onClick={ deleteIconClick(material) }>
            <Delete/>
          </IconButton>
          <IconButton onClick={ editIconClick(material) }>
            <Edit/>
          </IconButton>
        </ListItemSecondaryAction>
```

**Review by jarinykanen — CHANGES_REQUESTED** [review]

File: src/components/admin/reusables.tsx [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+  const fetchReusableMaterials = async () => {
+    if (!keycloak?.token) {
+      return;
+    }
+
+    setLoading(true);
+
+    try {
+      const fetchedMaterials = await Api.getReusableMaterialApi(keycloak.token).listReusableMaterials();
+      setMaterials(fetchedMaterials);
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:41]:
```suggestion
      setMaterials(await Api.getReusableMaterialApi(keycloak.token).listReusableMaterials());
```

File: src/components/admin/reusables.tsx [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+    setEditingMaterial(true);
+  };
+
+  /**
+   * Even handler for editable material name change
+   */
+  const handleEditableNameChange: React.ChangeEventHandler<HTMLTextAreaElement | HTMLInputElement> = ({ target }) => {
+    if (editableMaterial) {
+      setEditableMaterial({ ...editableMaterial, name: target.value });
+    }
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:147]:
```suggestion
    editableMaterial && setEditableMaterial({ ...editableMaterial, name: target.value });
```

File: src/components/admin/reusables.tsx [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+      title={ strings.adminScreen.deleteReusableMaterialDialog.title }
+      positiveButtonText={ strings.generic.confirm }
+      cancelButtonText={ strings.generic.cancel }
+    >
+      <Typography>
+        { strings.adminScreen.deleteReusableMaterialDialog.text }
+      </Typography>
+      <Typography variant="subtitle1">
+        { deletableMaterial?.name }
+      </Typography>
```
Comment (jarinykanen) [inline src/components/admin/reusables.tsx:215]: These could be combined with strings.formatString() function.

**Review by jarinykanen — CHANGES_REQUESTED** [review]

File: src/components/admin/reusables.tsx (hunk 12 above)
Comment (EVP1989) [reply src/components/admin/reusables.tsx]: Fixed

**Pull request merged**